### PR TITLE
Created decomposition.jl and added ear decompostion function

### PR DESCRIPTION
**Pull request by SaiHarsh**
All the necessary and required definitions and comments are available in code. 
Input: Any simple undirected unweighted connected graph with ne > nv otherwise it will return `[]`.
This algorithm will work on self loops too, so no need to remove them.

### CODE REVIEW

**Comment by simonschoelly** [comment]
There is something weird going on. Maybe your closures overwhelm Julia's type inference on older Julia versions. Try pulling out all your inner functions outside of the `ear_decomposition` function and see if it still fails.

**Comment by SaiHarsh** [comment]
Thanks, yes that was the problem.
I thought as the `DFS` and the `traversal` functions are specific to `ear_decomposition` it doesn't make sense to keep them as generic functions, so I decided to make them as sub-functions of `ear_decomposition` function.
The new changes requires no function declaration inside a function.
Can you have a look and give me your feedback, I tried to add all the necessary comments in it but may have missed few.

File: src/decomposition.jl [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+```
+
+### References
+Schmidt, Jens M. (2013a), "A Simple Test on 2-Vertex- and 2-Edge-Connectivity",
+Information Processing Letters, 113 (7): 241–244, arXiv:1209.0700, 
+doi:10.1016/j.ipl.2013.01.016.
+"""
+
+function ear_decomposition end
+@traitfn function ear_decomposition(g::AG::(!IsDirected), s::Integer) where {T, AG<:AbstractGraph{T}} !is_connected(g) && throw(ArgumentError("Graph must be connected"))
```
Comment (simonschoelly) [inline src/decomposition.jl]: I think you should put a line break before `!is_connected` otherwise it looks as if this test still belongs to the function signature.
Btw. is it necessary that the graph is connected? Or could the decomposition also be done on each connected component?

File: src/decomposition.jl [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+
+    # List to store the order in which dfs visits vertices.
+    dfs_order = []
```
Comment (simonschoelly) [inline src/decomposition.jl]: `[]` has type `Vector{Any}`, so this will be very slow. You probably should replace it by `T[]`

File: src/decomposition.jl [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+
+    # Boolean dict to mark vertices as visited or unvisited during
```
Comment (simonschoelly) [inline src/decomposition.jl]: Maybe dict is not the right word here, as one could confuse it with a dictionary.

File: src/decomposition.jl [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+    # Dfs traversal in graph.
+    seen = zeros(Bool, nv(g))
+
+    # Boolean dict to mark vertices as visited or unvisited in
+    # Dfs tree traversal.
+    traversed = Set()
```
Comment (simonschoelly) [inline src/decomposition.jl]: Maybe we should also use `zeros(Bool, nv(g))` here? Sets are kind of slow. And if it is a set, it should be `Set{T}()` because `Set()` produces a set of type `Set{Any}`.

File: src/decomposition.jl [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+
+    # Dict to store parent vertex of all the visited vertices.
+    parents = zeros(T, nv(g))
+
+    # List to store visit_time of vertices in Dfs traversal.
+    value = Dict()
```
Comment (simonschoelly) [inline src/decomposition.jl]: Also a vector might be faster here and `Dict()` has type `Dict{Any}`

File: src/decomposition.jl [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+
+    # List to store all the chains and cycles of the input graph G.
+    chains = []
```
Comment (simonschoelly) [inline src/decomposition.jl]: see above

File: src/decomposition.jl [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+
+    # Construct depth-first search tree from source s.
+    S = Vector{T}([s])
+
+    # mark source vertex s as seen
+    seen[s] = true
+
+    parents[s] = s
+    append!(dfs_order, s)
```
Comment (simonschoelly) [inline src/decomposition.jl]: `append!` is for multiple elements. Use `push!` instead.

File: src/decomposition.jl (hunk 2 above)
Comment (sbromberger) [reply src/decomposition.jl]: We prefer the long-form syntax for vector creation:
```
Vector{T}()
```

File: src/decomposition.jl (hunk 1 above)
Comment (SaiHarsh) [reply src/decomposition.jl]: Yes, I can apply `ear_decomposition` on each connected component, I will do some necessary changes for it.

File: src/decomposition.jl (hunk 2 above)
Comment (SaiHarsh) [reply src/decomposition.jl]: Okay I will keep as `Vector{T}()`

File: src/decomposition.jl (hunk 4 above)
Comment (SaiHarsh) [reply src/decomposition.jl]: Yes, `Set{T}()` will restrict to T vertices. I will updated it.

Comment (simonschoelly) [reply src/decomposition.jl]: Would it be a big waste of space, if we used a vector instead of a set here?

File: src/decomposition.jl (hunk 6 above)
Comment (SaiHarsh) [reply src/decomposition.jl]: This change will result in a continuous list, not a nested list for example
`chains = [] -> [1, 2, 3, 4, 5, 6]`
`chains = Vector{Vector{T}}() -> [[1, 2, 3], [4, 5, 6]]`

Correct if I am wrong.

File: src/decomposition.jl (hunk 7 above)
Comment (SaiHarsh) [reply src/decomposition.jl]: Yes, I will update it.

File: src/decomposition.jl (hunk 6 above)
Comment (simonschoelly) [reply src/decomposition.jl]: So right now, you have
```
chains = []
chain =  [ ... ]
push!(chains, chain)
= > chains == [ chain ] == [ [  ... ] ]
```
So you are also getting a nested list. Did you want to have a flat list instead?

Comment (SaiHarsh) [reply src/decomposition.jl]: I want nested list as each list will represent either a cycle or a chain

File: src/decomposition.jl (hunk 5 above)
Comment (SaiHarsh) [reply src/decomposition.jl]: Yes, I can replace `Dict()` with `zeros(T, nv(g))` as each node will get pop from the stack only once.

File: src/decomposition.jl (hunk 6 above)
Comment (simonschoelly) [reply src/decomposition.jl]: In that case,
```
chains = Vector{Vector{T}}()
```
seems to be the right thing to do.

File: src/decomposition.jl (hunk 4 above)
Comment (SaiHarsh) [reply src/decomposition.jl]: I am using `set()` only to see whether a node is a part of a chain or not, so I can replace it with a boolean vector. 
Thanks for pointing it out.

**Comment by sbromberger** [comment]
@simonschoelly in your opinion, is this ready for final review?

File: src/decomposition.jl [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+                value[u] = i
+            end
+
+            # Traverse all the non Tree edges(G-T (non-tree edges)), according to
+            # DFS order to find cycles and chains by traversing in DFS tree.
+            for u in dfs_order
+                for neighbor in neighbors(g, u)
+                    if (value[u] < value[neighbor] && u != parents[neighbor])
+
+                        # Make the firt end of non-tree edge visited
```
Comment (simonschoelly) [inline src/decomposition.jl]: first

File: src/decomposition.jl [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+
+An ear decomposition of an undirected graph `G` is a partition of its
+set of edges into a sequence of ears, such that the one or two endpoints
+of each ear belong to earlier ears in the sequence and such that the
+internal vertices of each ear do not belong to any earlier ear.
+
+For more information, see the 
+[Ear Decomposition](https://en.wikipedia.org/wiki/Ear_decomposition)
+
+It will throw an error if input graph is directed/disconnected graph.
```
Comment (simonschoelly) [inline src/decomposition.jl]: This sentence is not true anymore.

File: src/decomposition.jl [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+
+An ear of an undirected unweighted connected graph `G` is a path `P` where
+the two endpoints of the path may coincide (i.e., form a cycle), but where
+otherwise no repetition of edges or vertices is allowed, so every internal
+vertex of `P` has degree two in `P`.
+
+An ear decomposition of an undirected graph `G` is a partition of its
+set of edges into a sequence of ears, such that the one or two endpoints
+of each ear belong to earlier ears in the sequence and such that the
+internal vertices of each ear do not belong to any earlier ear.
```
Comment (simonschoelly) [inline src/decomposition.jl]: Isn't this sentence nearly exactly the one on Wikipedia? We might get into trouble if we copy such things, so you might have to write that in your own words.

File: src/decomposition.jl [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+```
+
+### References
+Schmidt, Jens M. (2013a), "A Simple Test on 2-Vertex- and 2-Edge-Connectivity",
+Information Processing Letters, 113 (7): 241–244, arXiv:1209.0700, 
+doi:10.1016/j.ipl.2013.01.016.
+"""
+
+function ear_decomposition end
+@traitfn function ear_decomposition(g::AG::(!IsDirected), s::Integer) where {T, AG<:AbstractGraph{T}}
```
Comment (simonschoelly) [inline src/decomposition.jl]: I think, we can give `s` the default value 1, as people might often not care, at what vertex we start.

File: src/decomposition.jl [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+
+    # Perform ear decomposition on each connected component of input graph.
+    for v in nodes
+        if !(seen[v])
+            # Start the depth first search from first vertex
+            # mark source vertex s as seen
+            seen[v] = true
+
+            # Initialize S and dfs_order for every new DFS_Tree
+            S = Vector{T}([v])
```
Comment (simonschoelly) [inline src/decomposition.jl]: Better write `S = T[v]`, as otherwise we first allocate a vector `[v]` and then copy that result to a second vector.

File: src/decomposition.jl [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+    # List to store all the chains and cycles of the input graph G.
+    chains = Vector{Vector{T}}()
+
+    # Construct depth-first search tree from source s.
+    S = Vector{T}([s])
+
+    nodes = Vector{T}()
+    push!(nodes, s)
+    for u in vertices(g)
+        push!(nodes, u)
```
Comment (simonschoelly) [inline src/decomposition.jl]: You can write `append!(nodes, vertices(g))` here. But maybe there is a way to use this algorithm, without having to create a vector of vertices?

File: src/decomposition.jl [hunk 14]
```diff
@@ -0,0 +1,167 @@
...
+            # Initialize S and dfs_order for every new DFS_Tree
+            S = Vector{T}([v])
+            dfs_order = T[]
+            parents[v] = v
+            push!(dfs_order, v)
+
+            while !isempty(S)
+                v = S[end]
+                u = 0
+                for n in neighbors(g, v)
```
Comment (simonschoelly) [inline src/decomposition.jl]: Can you use another symbol than `n` here for the neighbors? `n` might be confusing, because usually it denotes a quantity.

File: src/decomposition.jl [hunk 15]
```diff
@@ -0,0 +1,167 @@
...
+
+            # Initialize S and dfs_order for every new DFS_Tree
+            S = Vector{T}([v])
+            dfs_order = T[]
+            parents[v] = v
+            push!(dfs_order, v)
+
+            while !isempty(S)
+                v = S[end]
+                u = 0
```
Comment (simonschoelly) [inline src/decomposition.jl]: You should write  `u = T(0)` or `u = zero(T)` here. The problem is that otherwise `u` has first type `Int` and then later type `T`. This leads to type instability and therefore to worse performance.

File: src/decomposition.jl (hunk 10 above)
Comment (SaiHarsh) [reply src/decomposition.jl]: Yes, it's from Wikipedia, I thought writing own will not explain clearly as compared to Wikipedia.

Ear Decomposition: Partition the edge set of undirected graph `G` into a
sequence of paths(ears), such that only the end vertices of each path are
present in earlier paths. 

Does this make sense?

File: src/decomposition.jl (hunk 14 above)
Comment (SaiHarsh) [reply src/decomposition.jl]: for neighbor in neighbors(g, u)?

File: src/decomposition.jl (hunk 13 above)
Comment (SaiHarsh) [reply src/decomposition.jl]: I wanted to run dfs first from source `s` later from the unvisited vertices(in case of disconnected).
for that, I am pushing first s, later all the remaining vertices.

To exclude creating a vector of vertices, one way is to have 2 copies of dfs code, one for s and other for remaining. but it will disturb the readability of code. Could you suggest any other better way?

File: src/decomposition.jl (hunk 10 above)
Comment (simonschoelly) [reply src/decomposition.jl]: Shouldn't that then be "sequence of paths or cycles (ears)"?

File: src/decomposition.jl (hunk 9 above)
Comment (simonschoelly) [reply src/decomposition.jl]: I don't think you have to throw an error for directed graphs, that is handled automatically by `@traitfn function ear_decomposition(g::AG::(!IsDirected), s::Integer)`

File: src/decomposition.jl (hunk 13 above)
Comment (simonschoelly) [reply src/decomposition.jl]: Maybe a boolean flag, that is set to true in the beginning, that tells us, that we should start from the source, and as soon as we are done with this vertex, we set this it to false and use `vertices(g)` ?

File: src/decomposition.jl (hunk 14 above)
Comment (simonschoelly) [reply src/decomposition.jl]: Yes, that should work

File: src/decomposition.jl (hunk 13 above)
Comment (SaiHarsh) [reply src/decomposition.jl]: Instead of using boolean can I use 

`for v in (append!(T[s], vertices(g)))` ?

File: src/decomposition.jl [hunk 16]
```diff
@@ -0,0 +1,156 @@
...
+```
+
+### References
+Schmidt, Jens M. (2013a), "A Simple Test on 2-Vertex- and 2-Edge-Connectivity",
+Information Processing Letters, 113 (7): 241–244, arXiv:1209.0700, 
+doi:10.1016/j.ipl.2013.01.016.
+"""
+
+function ear_decomposition end
+@traitfn function ear_decomposition(g::AG::(!IsDirected), s::Int64=1) where {T, AG<:AbstractGraph{T}}
```
Comment (simonschoelly) [inline src/decomposition.jl]: Can we have either `s::Integer` or `s::T`, so that this function also works with other eltypes? I would prefer the former case, then you just have to convert `s` to a vertex of type `T` somewhere in your code.

**Comment by simonschoelly** [comment]
Can you add a test for a graph with another eltype than `Int` here? Also you probably should write some tests that cover the usual edge cases, such as graphs with zero and one vertex and graphs with a self-loop.

**Comment by sbromberger** [comment]
Hi,

Where are we on this PR?

**Comment by SaiHarsh** [comment]
Hey, 

Current update, I have added all the changes recommended by @simonschoelly , I feel it's ready for final review.

Please have a look and let me know if any more changes are needed.
